[Fix #317] Allow :sorted keyword in grouped back-ends and some doc improvements

vspinu opened the pull request:
I have designed it such that further addition of keywords will not break older versions of company. This is why all keywords are removed before the `pcase`.

vspinu commented:
BTW, I think `:with` keyword is not really suggestive of the task that it performs. How about `:prefix-ignore-rest`?

dgutov commented:
> How about :prefix-ignore-rest?

It makes more sense on the lower level, but misses the advantage of the more succinct `:with`: separating backends semantically in two groups, say, primary and secondary. It corresponds more closely to the effective meaning: "if none of the backends before :with are used, none of those after it are used either".

One has to read the docstring anyway to find out about the keyword's existence, so maybe it's okay if its meaning is not 100% apparent from its name, as long as they correlate well.

vspinu commented:
>  It corresponds more closely to the effective meaning: "if none of the backends before :with are used, none of those after it are used either".

Sorry, but :with doesn't ring a bell with me at all. The above sentence is imprecise; "not used" can mean anything. For a developer of a backend who  is reading the docstring it's crucial to understand what's going on a low level. You can ask for the opinion of others, but  I really had trouble understanding what :with meant till I got down to he code. 

The main problem is a semantic one.  It doesn't make sense for the "with" preposition to refer to the previous items. If anything :with in `(A B :with C D)` should refer to C and D. That is, "use prefix command *with* C and D" , which is exactly the opposite of what it does right now. 

You can have `:prefix-ignore` if length is a concern. `(A B :prefix-ignore C D)`.

dgutov commented:
> For a developer of a backend who is reading the docstring it's crucial to understand what's going on a low level.

But that a user-level keyword. A developer can read the implementation anyway.

> The above sentence is imprecise; "not used" can mean anything

Maybe; you can suggest improvements. What I mean it, skipped for the purpose of completion, isn't asked for its completions, etc. Which is a consequence of returning `nil` to `prefix`, but to make that connection, the user needs to read a different part of documentation.

> It doesn't make sense for the "with" preposition to refer to the previous items.

Why not? In English, "with" connects words that come before it with words that come after. "spoon with honey", for instance.

> You can ask for the opinion of others, but I really had trouble understanding what :with meant till I got down to he code.

I did ask for suggestions for the name of this keyword: #79.

vspinu commented:
>  "with" connects words that come before it with words that come after. "spoon with honey", 

I hope you don't mean that  `(A B :with C)`  is translated into "A, B with C".  `:with` doesn't connect A, B and C, but rather a computation of some sort (prefix filter) with some backends from the list. 

So your spoon is "prefix command" and your "honey" is backend C. So you get "prefix command works with C". And confusingly, prefix command currently does exactly the opposite - it works with A and B.

vspinu commented:
How about `:split`? It naturally refers to the list which is split in two. It can also be used in other contexts, not just with prefix filter (as it is currently happening with `company-sort-by-backend-importance`).

vspinu commented:
> How about :split? 

This might be great for multi-splits `(A B :split C :split D E)`. For instance `company-sort-by-backend-importance` can sort within each group of backens, but preserve the order between backends.

dgutov commented:
> I hope you don't mean that (A B :with C) is translated into "A, B with C". :with doesn't connect A, B and C

It does: it says "C is used only if A or B is used". And actually, C's `prefix` *is* consulted in this case: it's compared to A's or B's prefix (whichever was non-nil), and C is queries for completions only is they are equal. There's more context in the issue I've linked above; let's continue this discussion there, if you like.

> How about :split?

You'd have to define its semantics first. They seem to be different.

vspinu commented:
> It does: it says "C is used only if A or B is used". And actually, C's `prefix`
> *is* consulted in this case: it's compared to A's or B's prefix (whichever was
> non-nil), and C is queries for completions only is they are equal. There's more
> context in the issue I've linked above; let's continue this discussion there, if
> you like.

This is sooo complicated. Switching to that thread instead.

dgutov commented:
I've switched to the hyphen-less spelling across the project. Please rebase this branch on top of it, and I'll merge.

Also disregard my objections against `:sorted` without value. The implementation is easier this way, and the difference is not that important.

vspinu commented:
Sorry I haven't had a chance to look at it again :( Will fix this tomorrow.

> Also disregard my objections against :sorted without value

Are you sure? If other options will be added in the future, they will  likely need values. Or it's simply not worth spending too much time on this given the imminent obsoletion?

dgutov commented:
> Or it's simply not worth spending too much time on this given the imminent obsoletion?

This.

vspinu commented:
Should be good to go!

dgutov commented:
Thanks! :thumbsup:
